util/dao/src/lib.rs

```diff
@@ -1,48 +1,267 @@
+use numext_fixed_hash::H256;
 use std::cmp::max;
+use std::sync::Arc;
 
-pub fn calculate_dao_data(
-    parent_block_number: BlockNumber,
-    parent_block_epoch_ext: &EpochExt,
-    parent_dao_stats: &DaoStats,
-    secondary_epoch_reward: Capacity,
-) -> Result<(u64, Capacity), FailureError> {
-    let parent_c = Capacity::shannons(parent_dao_stats.accumulated_capacity);
-    let parent_g2 = calculate_g2(
-        parent_block_number,
-        parent_block_epoch_ext,
-        secondary_epoch_reward,
-    )?;
-    let parent_g = parent_block_epoch_ext
-        .block_reward(parent_block_number)?
-        .safe_add(parent_g2)?;
-    let current_c = parent_c.safe_add(parent_g)?;
-
-    let parent_ar = parent_dao_stats.accumulated_rate;
-    let current_ar = u128::from(parent_ar) * u128::from((parent_c.safe_add(parent_g2)?).as_u64())
-        / (max(u128::from(parent_c.as_u64()), 1));
-
-    Ok((current_ar as u64, current_c))
+pub struct DaoCalculator<'a, CS, DL> {
+    pub consensus: &'a Consensus,
+    pub store: Arc<CS>,
+    pub data_loader: DL,
 }
 
-pub fn calculate_maximum_withdraw(
-    output: &CellOutput,
-    deposit_dao_stats: &DaoStats,
-    withdraw_dao_stats: &DaoStats,
-) -> Result<Capacity, FailureError> {
-    let occupied_capacity = output.occupied_capacity()?;
-    let counted_capacity = output.capacity.safe_sub(occupied_capacity)?;
+impl<'a, CS: ChainStore> DaoCalculator<'a, CS, DataLoaderWrapper<CS>> {
+    pub fn new(consensus: &'a Consensus, store: Arc<CS>) -> Self {
+        let data_loader = DataLoaderWrapper::new(Arc::clone(&store));
+        DaoCalculator {
+            consensus,
+            store,
+            data_loader,
+        }
+    }
+
+    pub fn primary_block_reward(&self, target: &Header) -> Result<Capacity, FailureError> {
+        let target_epoch = self
+            .store
+            .get_block_epoch_index(target.hash())
+            .and_then(|index| self.store.get_epoch_ext(&index))
+            .ok_or(Error::InvalidHeader)?;
+
+        target_epoch.block_reward(target.number())
+    }
+
+    pub fn secondary_block_reward(&self, target: &Header) -> Result<Capacity, FailureError> {
+        let target_parent_hash = target.parent_hash();
+        let target_parent = self
+            .store
+            .get_block_header(target_parent_hash)
+            .ok_or(Error::InvalidHeader)?;
+        let target_epoch = self
+            .store
+            .get_block_epoch_index(target.hash())
+            .and_then(|index| self.store.get_epoch_ext(&index))
+            .ok_or(Error::InvalidHeader)?;
+
+        let target_g2 = calculate_g2(
+            target.number(),
+            &target_epoch,
+            self.consensus.secondary_epoch_reward(),
+        )?;
+        let (_, _, target_parent_u) = extract_dao_data(target_parent.dao())?;
+        let (_, target_c, _) = extract_dao_data(target.dao())?;
+        let reward = u128::from(target_g2.as_u64()) * u128::from(target_parent_u.as_u64())
```

Comment:
If we save `U_prev` in the block header, we don't need to get parent header to calculate the miner reward. And we also get following benefits:

- The `dao` in genesis block is determined, we don't have to calculate it on the fly.
- If miners want's to pick transactions based on the block assembler template, they don't need to update `dao` in the header, since it is affected by the transactions commit in the current block.

Reply:
At first I thought this was a good idea, but after some thought I feel it's gonna be quite weird to save `U_prev` in the block header. The problem is that in order to calculate `U_prev`, we need to resolve the whole previous block(since we need to calculate sum of all input capacities consumed in previous block), which is gonna be quite strange. I feel like to validate a block, we should only rely on the resolved data of the current block, not previous block.

And yes miner needs to update `dao` when picking transactions in the header, notice they will also need to calculate very similar things such as fees they can get, and the process of calculating `U` in `dao` resembles fee calculation a lot, so I doubt it will bring much effort here.